source/CustomSale.cpp

```diff
@@ -0,0 +1,333 @@
+#include "DataNode.h"
+#include "GameData.h"
+#include "Planet.h"
+#include "Set.h"
+
+#include <string>
+
+using namespace std;
+
+namespace 
+{
+	const auto show = map<CustomSale::SellType, const string> {
+		{CustomSale::SellType::NONE, ""},
+		{CustomSale::SellType::VISIBLE, ""},
+		{CustomSale::SellType::IMPORT, "import"},
+		{CustomSale::SellType::HIDDEN, "hidden"},
+	};
+	const double DEFAULT = -100000.;
+}
+
+
+
+void CustomSale::Load(const DataNode &node, const Set<Sale<Outfit>> &items, const Set<Outfit> &outfits, const string &mode)
+{
+	for(const DataNode &child : node)
+	{
+		const string &token = child.Token(0);
+		bool isValue = (child.Token(0) == "value");
+		bool isOffset = (child.Token(0) == "offset");
+		if((token == "clear" || token == "remove"))
+		{
+			if(child.Size() == 1)
+				Clear();
+			else if(child.Token(1) == "outfit")
+			{
+				if(child.Size() >= 3)
+				{
+					const Outfit *outfit = outfits.Get(child.Token(2));
+					relativeOutfitPrices.erase(outfit);
+					relativeOutfitOffsets.erase(outfit);
+				}
+				else
+				{
+					relativeOutfitOffsets.clear();
+					relativeOutfitPrices.clear();
+				}
+			}
+			else if(child.Token(1) == "outfitter")
+			{
+				if(child.Size() >= 3)
+				{
+					const Sale<Outfit> *outfitter = items.Get(child.Token(2));
+					relativePrices.erase(outfitter);
+					relativeOffsets.erase(outfitter);
+				}
+				else
+				{
+					relativeOffsets.clear();
+					relativePrices.clear();
+				}
+			}
+			else if(child.Token(1) == "location")
+				locationFilter = LocationFilter{};
+			else if(child.Token(1) == "conditions")
+				conditions = ConditionSet{};
+			else
+				child.PrintTrace("Skipping unrecognized clearing/deleting:");
+		}
+		else if(token == "add")
+		{
+			if(child.Token(1) == "location" && child.Size() == 1)
+				locationFilter.Load(child);
+			else if(child.Token(1) == "conditions")
+				conditions.Load(child);
+			else
+				child.PrintTrace("Skipping unrecognized add:");
+		}
+		else if(token == "hidden")
+			sellType = SellType::HIDDEN;
+		else if(token == "import")
+			sellType = SellType::IMPORT;
+		else if(token == "visible")
+			sellType = SellType::VISIBLE;
+		else if(token == "location")
+		{
+			if(child.Size() == 1)
+			{
+				location = nullptr;
+				locationFilter = LocationFilter{};
+				locationFilter.Load(child);
+			}
+			else if(child.Size() == 2)
+			{
+				location = GameData::Planets().Get(child.Token(1));
+				locationFilter = LocationFilter{};
+				if(child.HasChildren())
+					child.PrintTrace("Warning: location filter ignored due to use of explicit planet:");
+			}
+			else
+				child.PrintTrace("Warning: use a location filter to choose from multiple planets:");
+		}
+		else if(token == "conditions")
+		{
+			conditions = ConditionSet{};
+			conditions.Load(child);
+		}
+		else if(mode == "outfits")
+		{
+			bool isAdd = false;
+			const Outfit *outfit = nullptr;
+			auto parseValueOrOffset = [isAdd, outfit](double &amount, const DataNode &line) {
```

Comment:
```suggestion
			auto parseValueOrOffset = [&isAdd, &outfit](double &amount, const DataNode &line) {
```
The type of variable capture is very important. The default is by-value capture, which means you were always invoking with uninitialized data (which is why the macOS checks were complaining).

I suggest adding an integration test that loads a test pilot and a custom sale definition in the test config, and visit the outfitter in that test case. @petervdmeer can help you get it sorted.

Reply:
Very good idea, I tested it manually way too many times, and only discovered recently such extensive automated tests were even possible

Reply:
Its unfortunately not possible have a custom sale definition in the test config, so these tests will have to wait as I explain below.